mypy/checkexpr.py

```diff
@@ -3654,6 +3650,29 @@ def visit_type_application(self, tapp: TypeApplication) -> Type:
             return AnyType(TypeOfAny.from_another_any, source_any=tp)
         return AnyType(TypeOfAny.special_form)
 
+    def apply_type_arguments_to_type_alias(
+        self, item: Type, tapp: TypeApplication
+    ) -> Tuple[bool, Type]:
+        item = get_proper_type(item)
+
+        if isinstance(item, Instance):  # TODO: what about other types?
+            tp = type_object_type(item.type, self.named_type)
+            return False, self.apply_type_arguments_to_callable(tp, item.args, tapp)
+        elif isinstance(item, (NoneType, TupleType, CallableType)):
+            return False, item  # We don't do anything for these types, they are ready
```

Comment:
This is not right, this should be a type of type alias in runtime context, not the target type itself (i.e. like `C` when appears in runtime context becomes ~`Type[C]`). In general, I think it is a bad idea to support arbitrary aliases in runtime context.

Btw one of my other PRs handle tuple/typeddict cases correctly (which will make sense when generic tuple types are enabled)

Reply:
What actually can be done, is to make the error message less cryptic. Just to explain "type application" is a thing where you explicitly apply type arguments to e.g. a generic function or generic class object. Python however doesn't support this for functions yet. Example:
```python
class C(Generic[T]):
    def __init__(self, x: T) -> None: ...

C[int]("no way")  # Error
```
IOW you can't just put arbitrary type expression in runtime context and expect it to do something meaningful (like what would `List[int] + List[int]` mean). So the error message should explain the user, that types should only live in type context (like type annotations or etc.) I am however not sure how to phrase this concisely.

Reply:
Yet another option (which will be consistent with how we treat unsubscripted aliases in runtime context) is to simply give all aliases, that don't have a meaning of a callable, type `object`. So this will be valid:
```python
def print_annotation(ann: object) -> None: ...

def test_print_optional() -> None:
    O = Optional[T]
    print_annotation(O[int])
    print_annotation(O[str])
```

I think this will be the best solution for now.